src/core/ast.ml

```diff
@@ -222,13 +223,23 @@ and refresh_computation vars = function
       Out (op, refresh_expression vars expr, refresh_computation vars comp)
   | In (op, expr, comp) ->
       In (op, refresh_expression vars expr, refresh_computation vars comp)
-  | Promise (op, abs, p, comp) ->
+  | Promise (None, op, abs, p, comp) ->
       let p' = Variable.refresh p in
       Promise
-        ( op,
+        ( None,
+          op,
           refresh_abstraction vars abs,
           p',
           refresh_computation ((p, p') :: vars) comp )
+  | Promise (Some k, op, abs, p, comp) ->
```

Comment:
It would be good to merge the two cases with something along the lines of
```
    let k', vars' =
      match k with
      | None -> (None, vars)
      | Some k -> let k' = refresh k in (k', (k, k') :: vars)
```